### ISSUE TITLE
RFC: read-only filesystem support

### ISSUE THREAD

**Issue by dkarlovi** [pimcore/pimcore#8531]
# Feature Request

## Feature description

In an increasingly cloud-native world, having an app think it must have its permanent storage directly attached is quite outdated when runtimes are ephemeral (like Docker containers or lamba functions) and the expectation is the [apps are stateless](https://12factor.net/processes).

Pimcore should support read-only local filesystem, meaning it should be able to (reasonably) work without being able to write to disk all the time. Symfony supports 

## Blockers

And their possible solutions.

1. **Assets** should be fully managed by a cloud storage solution like S3 or Azure Blob Storage. I know about the [S3 hack](https://pimcore.com/docs/pimcore/current/Development_Documentation/Installation_and_Upgrade/System_Setup_and_Hosting/Amazon_AWS_Setup/Amazon_AWS_S3_Setup.html), but this is just a band-aid, asset-related filesystem interactions should be done via something like [Flysystem](https://github.com/thephpleague/flysystem), allowing for full flexibility when choosing the asset storage.
2. **Logs** should be written to STDOUT / STDERR, should already be possible since Monolog is used and can be reconfigured.
3. For **Cache**, Symfony differentiates several concepts, "system cache" and "app cache". The system cache is basically prebuilt Symfony container and it should already be read-only, [it's also called "build dir" in Symfony 5.2+](https://github.com/symfony/symfony/pull/36515), this should be prebuilt in the Docker image before deploying anyway. App cache is basically application data and it's already possible to use some shared cache like Redis, Memcached, the database, etc.
4. **Sessions** are already handled by Redis, Memcached, database
5. **Class definitions, object bricks and similar**, it should just be treated as read-only anyway. In all my use cases, I don't want these to be updated on production anyway, they're only deployed there after being done by the developers, versioned and tested, etc. I should be able to deploy my changes though. In the admin they get treated like read-only, maybe a note and all the actions are no-op. **Pimcore should tell its extensions it's in read-only mode** so they too know to disable their interactions where appropriate (example, DataDefinitions disables updating definitions).
6. `var/config/system.yml` should also be stored in persistent storage, why not in the database?
7. any other stuff would actually be some temporary cache, it's perfectly valid to use `/tmp` or the available equivalent for that, but the expectation is it could be gone the next time you run it

Basically, any time Pimcore tries to store anything on the filesystem, it should actually think:
1. is this a asset file? In which case Flysystem
2. is this data structure? in which case it's read-only in production, it should be deployed.
3. is this logging? In which case STDOUT / STDERR
4. is this system cache? in which case it's read-only in production, it should be prebuilt.
5. is this application cache? in which case it should go to the shared cache
6. is this application config? In which case it should be in the database
7. is it something else? In which case use `/tmp` but expect it to disappear

**Comment by dpfaffenbauer** [comment]
2. is possible, monolog/symfony handle that perfectly
 5. absolutely agree here, links somewhat to https://github.com/pimcore/pimcore/issues/8256, https://github.com/pimcore/pimcore/issues/8255 and https://github.com/pimcore/pimcore/issues/8250
 6. Not everything honestly. Things like the language with fallback should be a database setting. debug_mode not sure, mail config is a symfony config anyway. so there is a bit of mix and match of database related configs (like languages) and env/symfony config based configurations (like mailer). IMO it should not be even necessary to need the system config UI in backend anymore.

**Comment by solverat** [comment]
6: https://github.com/pimcore/pimcore/issues/8156

**Comment by brusch** [comment]
Most of the mentioned stuff should be now covered by Pimcore 10. 
If you think there's something missing please create a seperate issue for each topic. 
Thanks a lot 👍
